java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java

```diff
@@ -671,7 +670,7 @@ void executeInstruction(Instruction instruction) {
         pop = popStack(1, instruction.opcode);
         sv = pop.values.get(0);
         TypedConstraint typedConstraint = programState.getConstraint(sv, TypedConstraint.class);
-        Type type = typedConstraint != null ? typedConstraint.type : Symbols.unknownType;
+        Type type = typedConstraint != null ? typedConstraint.getType(semanticModel) : Symbols.unknownType;
```

Comment:
this means we will still keep reference to `Type` inside exceptional symbolic value, which will be stored in yield and thus shared in the behavior cache. It seems to me that we need to do same change in `ExceptionalYield`

Reply:
see second commit.